### PR TITLE
Convert SlimSAM checkpoints

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR extends the conversion script of SAM (Segment Anything) to also support [SlimSAM](https://github.com/czg1225/SlimSAM/tree/master) checkpoints.

SlimSAM is a compressed (pruned) version of SAM, claims to outperform FastSAM and MobileSAM. Looks cool! Below are the SAM-vit-base (top) versus the SlimSAM-77-uniform (bottom) results.

Screenshot 2024-01-07 at 20 59 40

see https://github.com/czg1225/SlimSAM/issues/4

### CODE REVIEW

File: src/transformers/models/sam/convert_sam_original_to_hf_format.py [hunk 1]
```diff
@@ -88,73 +133,71 @@ def replace_keys(state_dict):
...
-            hidden_size=1280,
-            num_hidden_layers=32,
-            num_attention_heads=16,
-            global_attn_indexes=[7, 15, 23, 31],
-        )
+def convert_sam_checkpoint(model_name, pytorch_dump_folder, push_to_hub):
+    if model_name == "slimsam-50-uniform":
+        checkpoint_path = "/Users/nielsrogge/Documents/SlimSAM/SlimSAM-50-uniform.pth"
+    elif model_name == "slimsam-77-uniform":
+        checkpoint_path = "/Users/nielsrogge/Documents/SlimSAM/SlimSAM-77-uniform.pth"
```
Comment (ArthurZucker) [inline src/transformers/models/sam/convert_sam_original_to_hf_format.py:140]: that's not very general

File: src/transformers/models/sam/convert_sam_original_to_hf_format.py [hunk 2]
```diff
@@ -88,73 +133,71 @@ def replace_keys(state_dict):
...
+
+        with torch.no_grad():
+            output = hf_model(**inputs)
+            scores = output.iou_scores.squeeze()
+
+        print(scores)
+        pass
+
+    elif model_name == "sam_vit_h_4b8939":
+        # assert scores[-1].item() == 0.579890251159668
```
Comment (ArthurZucker) [inline src/transformers/models/sam/convert_sam_original_to_hf_format.py:184]:
```suggestion
```

File: src/transformers/models/sam/convert_sam_original_to_hf_format.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/models/sam/convert_sam_original_to_hf_format.py:140]: let' s add a argument parsed for the path wdyt?

**Review by ArthurZucker — APPROVED** [review]
Left a few nits but LGTM otherwise

File: src/transformers/models/sam/convert_sam_original_to_hf_format.py (hunk 1 above)
Comment (NielsRogge) [reply src/transformers/models/sam/convert_sam_original_to_hf_format.py:140]: Yes good idea

File: src/transformers/models/sam/convert_sam_original_to_hf_format.py (hunk 2 above)
Comment (NielsRogge) [reply src/transformers/models/sam/convert_sam_original_to_hf_format.py:184]: Hi I think it's a bad practice to remove assertion checks from conversion scripts.

e.g. for llava 1.5 or SAM, there's no proof that we match the original logits (and I wasn't also able to match them - for llava 1.5 it turned out to be because of the image which wasn't padded as done in the original repo)

Hence I'm advocating to keep assertions in conversion scripts (as they are mostly run by us)
